source/particles/particle_handler.cc

```diff
@@ -205,66 +227,158 @@ namespace Particles
 
   template <int dim, int spacedim>
   void
-  ParticleHandler<dim, spacedim>::update_cached_numbers()
+  ParticleHandler<dim, spacedim>::reset_particle_container(
+    const Triangulation<dim, spacedim> *triangulation,
+    particle_container &                given_particles)
   {
-    number_of_locally_owned_particles                  = 0;
-    types::particle_index local_max_particle_index     = 0;
-    types::particle_index local_max_particles_per_cell = 0;
+    given_particles.clear();
+    for (unsigned int i = 0; i < 3; ++i)
+      given_particles.emplace_back(
+        std::vector<typename PropertyPool<dim, spacedim>::Handle>(),
+        triangulation->end());
```

Comment:
If triangulation is a `nullptr` this line dereferences a null pointer (which is undefined behavior), which sometimes succeeds (as apparently in the deal.II tests) and sometimes is detected as a corruption of the stack (as on some of my machines for the ASPECT tests). We should only execute these steps if a triangulation is set, because if it is not set the particle handler is not in a usable state anyways, and `initialize` or `copy_from` have to be called before it is usable. Maybe also not call this function at all in the default constructor, because it will be a no-op?
Also I have not seen a reason to give triangulation as an argument to this function, it will always be the stored triangulation pointer (and that is at least a smart pointer), do you think it makes sense to remove the argument? The comparison would then probably need to be `if (triangulation != SmartPointer<const Triangulation<dim, spacedim>, ParticleHandler<dim, spacedim>>())` because we cannot dereference an uninitialized smart pointer (I havent tried this).

This was one of the more interesting debugging sessions :-)

Reply:
Thanks a lot for the help, this has indeed been wrong. I removed the argument `triangulation` and construct the invalid iterator there. I still kept support for `triangulation == nullptr`, as I want `ParticleHandler::begin()` to return a valid (empty) range rather than throw an exception with internal error. If we want, we could of course change this again and safe three lines of code.

Reply:
That is a good point, and yes the current version is fine. That was the last of my comments, ready from my side.